### PR TITLE
Grid system

### PR DESCRIPTION
**Pull request by Will1400**
New grid system
- Uses cells to store data for a specific point
- Does not use colliders for placement

GridController
- Has a grid for the map and a seperate grid for towers

GridObstacle
- Is loaded by GridController on map load, and is added to the map grid as a GameObject Cell with owner of the obstacle

### CODE REVIEW

File: Assets/Scripts/Systems/Grid/Cell/Base/BaseCell.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿namespace TDGame.Systems.Grid.Cell.Base
+{
+    public abstract class BaseCell
+    {
+        
+    }
+}
```
Comment (75nul38) [inline Assets/Scripts/Systems/Grid/Cell/Base/BaseCell.cs:7]:
```suggestion
}

```

File: Assets/Scripts/Systems/Grid/Cell/Implementations/EmptyCell.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+﻿using TDGame.Systems.Grid.Cell.Base;
+
+namespace TDGame.Systems.Grid.Cell.Implementations
+{
+    public class EmptyCell : BaseCell
+    {
+    }
+}
```
Comment (75nul38) [inline Assets/Scripts/Systems/Grid/Cell/Implementations/EmptyCell.cs:8]:
```suggestion
}

```

File: Assets/Scripts/Systems/Grid/Cell/Implementations/GameObjectCell.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    public class GameObjectCell : BaseCell, IDisplayCell
+    {
+        public GameObject Owner;
+
+        public override string ToString()
+        {
+            return Owner.ToString();
+        }
+    }
+}
```
Comment (75nul38) [inline Assets/Scripts/Systems/Grid/Cell/Implementations/GameObjectCell.cs:16]:
```suggestion
}

```

File: Assets/Scripts/Systems/Grid/Cell/Interfaces/IDisplayCell.cs [hunk 4]
```diff
@@ -0,0 +1,6 @@
+﻿namespace TDGame.Systems.Grid.Cell.Interfaces
+{
+    public interface IDisplayCell
+    {
+    }
+}
```
Comment (75nul38) [inline Assets/Scripts/Systems/Grid/Cell/Interfaces/IDisplayCell.cs:6]:
```suggestion
}

```

File: Assets/Scripts/Systems/Grid/Data/GridArea.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            return new Vector3(position.x, 0, position.y) * .5f;
+        }
+
+        public Vector2 ConvertToWorldSize()
+        {
+            return new Vector2(width, height) * .5f;
+        }
+    }
+}
```
Comment (75nul38) [inline Assets/Scripts/Systems/Grid/Data/GridArea.cs:37]:
```suggestion
}

```

File: Assets/Scripts/Systems/Grid/Grid2D.cs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            return y * sizeX + x;
+        }
+
+        public void Dispose()
+        {
+            grid = null;
+        }
+    }
+}
```
Comment (75nul38) [inline Assets/Scripts/Systems/Grid/Grid2D.cs:138]:
```suggestion
}

```

File: Assets/Scripts/Systems/Grid/InGame/GridAreaController.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        void OnDrawGizmosSelected()
+        {
+            var worldPosition = originPoint.position;
+            var size = area.ConvertToWorldSize();
+            Gizmos.color = Color.magenta;
+            Gizmos.DrawLine(worldPosition, worldPosition + new Vector3(size.x, 0, 0));
+            Gizmos.DrawLine(worldPosition, worldPosition + new Vector3(0, 0, size.y));
+        }
+    }
+}
```
Comment (75nul38) [inline Assets/Scripts/Systems/Grid/InGame/GridAreaController.cs:30]:
```suggestion
}

```

File: Assets/Scripts/Systems/Grid/InGame/GridObstacle.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        void OnDrawGizmosSelected()
+        {
+            var worldPosition = originPoint.position;
+            var size = area.ConvertToWorldSize();
+            Gizmos.color = Color.magenta;
+            Gizmos.DrawLine(worldPosition, worldPosition + new Vector3(size.x, 0, 0));
+            Gizmos.DrawLine(worldPosition, worldPosition + new Vector3(0, 0, size.y));
+        }
+    }
+}
```
Comment (75nul38) [inline Assets/Scripts/Systems/Grid/InGame/GridObstacle.cs:29]:
```suggestion
}

```

File: Assets/Scripts/Systems/Grid/InGame/GridController.cs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+            Debug.DrawLine(mapGrid.GridToWorldPosition(0, (int) gridSize.y),
+                mapGrid.GridToWorldPosition((int) gridSize.x, (int) gridSize.y),
+                Color.white, 100f);
+            Debug.DrawLine(mapGrid.GridToWorldPosition((int) gridSize.x, 0),
+                mapGrid.GridToWorldPosition((int) gridSize.x, (int) gridSize.y),
+                Color.white, 100f);
+        }
+    }
+}
```
Comment (75nul38) [inline Assets/Scripts/Systems/Grid/InGame/GridController.cs:113]:
```suggestion
}

```

File: Assets/Tests/Systems/Grid/TestGridCellMethods.cs [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+        [TestCase(2, 4, true)]
+        [TestCase(5, 15, true)]
+        [TestCase(1, 40, false)]
+        public void TestGridSetCell(int x, int y, bool expected)
+        {
+            var grid = new Grid2D(21, 32);
+            Assert.That(grid.SetCell(x, y, new EmptyCell()), Is.EqualTo(expected));
+        }
+    }
+}
```
Comment (75nul38) [inline Assets/Tests/Systems/Grid/TestGridCellMethods.cs:19]:
```suggestion
}

```

File: Assets/Tests/Systems/Grid/TestGridConversion.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        public void TestWorldToGridPosition(float cellSize, float worldX, float worldY, float worldZ, int expectedX,
+            int expectedY
+        )
+        {
+            var grid = new Grid2D(10, 20, cellSize);
+            var converted = grid.WorldToGridPosition(new Vector3(worldX, worldY, worldZ));
+            Assert.That(new Vector2Int(expectedX, expectedY), Is.EqualTo(converted));
+        }
+    }
+}
```
Comment (75nul38) [inline Assets/Tests/Systems/Grid/TestGridConversion.cs:33]:
```suggestion
}

```

File: Assets/Tests/Systems/Grid/TestGridInitialization.cs [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        [TestCase(2, 1, 0.5f)]
+        public void TestFullConstructor(int x, int y, float cellSize)
+        {
+            Grid2D grid = new Grid2D(x, y, cellSize);
+
+            Assert.That(x * y, Is.EqualTo(grid.grid.Length));
+            Assert.That(cellSize, Is.EqualTo(grid.CellSize));
+        }
+    }
+}
```
Comment (75nul38) [inline Assets/Tests/Systems/Grid/TestGridInitialization.cs:37]:
```suggestion
}

```

File: Assets/Tests/Systems/Grid/TestGridUtility.cs [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+        [TestCase(20, 13, 20, 13, false)]
+        [TestCase(200, 13, 123, 12, true)]
+        [TestCase(120, 120, 0, 0, true)]
+        public void TestGridIsValidPoint(int sizeX, int sizeY, int x, int y, bool expected)
+        {
+            var grid = new Grid2D(sizeX, sizeY);
+            Assert.That(expected, Is.EqualTo(grid.IsValidGridPosition(x, y)));
+        }
+    }
+}
```
Comment (75nul38) [inline Assets/Tests/Systems/Grid/TestGridUtility.cs:29]:
```suggestion
}

```

**Review by 75nul38 — APPROVED** [review]

File: Assets/Tests/Systems/Grid/TestGridUtility.cs (hunk 13 above)
Comment (Will1400) [reply Assets/Tests/Systems/Grid/TestGridUtility.cs:29]: Empty line ending is not used in other places, so I will not add them here

**Pull request merged**